sound/soc/amd/acp/amd-sdw-acpi.c

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license. When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2023 Advanced Micro Devices, Inc. All rights reserved.
+//
+// Authors: Vijendar Mukunda <Vijendar.Mukunda@amd.com>
+
+/*
+ * SDW AMD ACPI scan helper function
+ */
+
+#include <linux/acpi.h>
+#include <linux/bits.h>
+#include <linux/bitfield.h>
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/export.h>
+#include <linux/fwnode.h>
+#include <linux/module.h>
+#include <linux/soundwire/sdw_amd.h>
+#include <linux/string.h>
+
+int amd_sdw_scan_controller(struct sdw_amd_acpi_info *info)
+{
+	struct acpi_device *adev = acpi_fetch_acpi_dev(info->handle);
+	u32 sdw_bitmap = 0;
+	u8 count = 0;
+	int ret;
+
+	if (!adev)
+		return -EINVAL;
+
+	/* Found controller, find links supported */
+	ret = fwnode_property_read_u32_array(acpi_fwnode_handle(adev),
+					     "mipi-sdw-manager-list", &sdw_bitmap, 1);
+	if (ret) {
+		dev_err(&adev->dev,
+			"Failed to read mipi-sdw-manager-list: %d\n", ret);
+		return -EINVAL;
+	}
+	count = hweight32(sdw_bitmap);
+	/* Check count is within bounds */
+	if (count > info->count) {
+		dev_err(&adev->dev, "Manager count %d exceeds max %d\n",
+			count, info->count);
+		return -EINVAL;
+	}
+
+	if (!count) {
+		dev_dbg(&adev->dev, "No SoundWire Managers detected\n");
+		return -EINVAL;
+	}
+	dev_dbg(&adev->dev, "ACPI reports %d SoundWire Manager devices\n", count);
+	info->link_mask = sdw_bitmap;
+	return 0;
+}
+EXPORT_SYMBOL_NS(amd_sdw_scan_controller, SND_AMD_SOUNDWIRE_ACPI);
```

Comment:
this is mostly similar to the code in sound/hda/intel-sdw-acpi.c

Probably best to try and share the code?

Reply:
1. Currently Intel following MIPI Disco Spec 1.0 which and ""mipi-sdw-master-count" property used to retrieve link count where as for AMD platforms, we are following MIPI Disco Spec 2.0 peripheral and we are referring "mipi-sdw-manager-list" to retrieve manager link mask. 
2. Intel has fixed max manager count where for AMD platforms it may vary. Thats why we are reading the max link count from parent private data structure.
3. We defined separate scanner API so that in future , if some extra variables need to be added for AMD platforms, we want to leave a place holder.

Still we can go for a common code, if Intel migrates to latest DisCo spec and keep checking minimal code "mipi-sdw-manager-list" in the scan controller API.

Reply:
yeah we're still to add the "mipi-sdw-manager-list", we've not seen it so far.

Reply:
As of today as scan functions are different for AMD & Intel platforms. Lets go with separate API.